lib/core/src/Cardano/Wallet/Registry.hs

```diff
@@ -137,6 +136,16 @@ insert
 insert (WorkerRegistry mvar) wrk =
     modifyMVar_ mvar (pure . Map.insert (workerId wrk) wrk)
 
+-- | Unregister a worker from the registry, but don't cancel the running task.
+--
+unregister
+    :: Ord key
+    => WorkerRegistry key resource
+    -> key
+    -> IO ()
+unregister (WorkerRegistry mvar) k =
+    modifyMVar_ mvar (pure . Map.delete k)
```

Comment:
Having both `unregister` and `remove` makes the module API somewhat non-orthogonal. What would you think of revising a bit the interface here by: 

- Keeping only one of the two, likely (`unregister` as it sounds more dual to `register`)
- Having `unregister` returns the actual `Worker key resource` when removed from the registry 
- Let the caller of `unregister` kill the worker thread itself

So, on call sites, instead of:

```hs
    liftHandler $ withWorkerCtx @_ @s @k ctx wid throwE $ \_ -> pure ()
    liftIO $ Registry.remove re wid
```

We would have:

```hs
    liftIO $ Registry.unregister re wid >>= \case
        Nothing  -> liftHandler $ throwE $ ErrNoSuchWallet wid
        Just wrk -> liftIO $ killThread $ workerThread wrk
```

which actually feels way less awkward that the artificial 'withWorkerCtx' call and make the whole thing much more explicit.

Reply:
@KtorZ I like the general idea, though I have one suggestion.

Since the registry is responsible for handling thread creation, perhaps it should also be responsible for thread destruction?

How about the following (names can be changed):
```hs
unregister
    :: Ord key
    => WorkerRegistry key resource
    -> key
    -> IO (Maybe (Worker key resource))   
unregister (WorkerRegistry mvar) k = do   
    mWorker <- Map.lookup k <$> readMVar mvar 
    modifyMVar_ mvar (pure . Map.delete k)
    pure mWorker 

terminate
    :: Ord key
    => WorkerRegistry key resource
    -> key
    -> IO ()  
terminate registry k =
    unregister registry k >>= traverse_ (killThread . workerThread)
```

Then, `deleteWallet` can be:
```hs
deleteWallet ctx (ApiT wid) = 
   liftHandler $ withWorkerCtx @_ @s @k ctx wid throwE $ \_ -> do 
      liftIO $ Registry.terminate re wid  
      liftIO $ removeDatabase df wid  
      return NoContent   
```

Reply:
@KtorZ 

I've revised the interface of `Registry` so that `register` and `unregister` are **dual** operations of one another.

* `register` is responsible for:
    1. adding an entry into the registry map.
    2. thread creation.
* `unregister` is responsible for:
    1. removing an entry from the registry map.
    2. thread destruction.

See commit b3865746ca0d85d0dbf5eee81bdd03e0c87340d1.